### ISSUE TITLE
Hang during session break

### ISSUE THREAD

**Issue by mykolap** [opentelecoms-org/jsmpp#6]
When I stop my SMPP simulator - session hangs
After about 15 min - computer hangs and only hard reset of PC can help

I watched logs and noticed that there are messages "AbstractSession.close() called",
but no "AbstractSession.close() done" messages in log.

The cause of hangs - enquireLinkSender.join();

I remove this part from org/jsmpp/session/AbstractSession.java and now all OK.

```
    // Make sure the enquireLinkThread doesn't wait for itself
    if (Thread.currentThread() != enquireLinkSender) {
        if (enquireLinkSender != null) {
            while(enquireLinkSender.isAlive()) {
                try {
                    enquireLinkSender.join();
                } catch (InterruptedException e) {
                    logger.warn("interrupted while waiting for enquireLinkSender thread to exit");
                }
            }
        }
    }
```

**Comment by pmoerenhout** [comment]
Hi Mykola, I tried to reproduce the issue with the org.jsmpp.examples.SMPPServerSimulator program and the OpenAndBindExample as client. But the thread join always succeed, and the client can always close normally. How do you reproduce the issue?

**Comment by mykolap** [comment]
We use jsmpp 2.1.0 from google before, but when we moved from java 6 to
java 8 - 2.1.0 stopped working.
I change jsmpp version to 2.2.1, but during tests with about 5 sessions in
async mode (3-10 async pdu per session on tx and 3-10 pdu per session on rx
on every session) I got a lot of deadlocks.
After 30 minutes of testing server stopped responding and only hard reset
helped.
After analyzing logs I understood that waiting on that piece of code caused
that locks.
After removing this code all errors disappeared and working fine during
last few monthes.


On Thu, Aug 20, 2015 at 4:07 PM, Pim Moerenhout <notifications@github.com>
wrote:

> Hi Mykola, I tried to reproduce the issue with the
> org.jsmpp.examples.SMPPServerSimulator program and the OpenAndBindExample
> as client. But the thread join always succeed, and the client can always
> close normally. How do you reproduce the issue?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opentelecoms-org/jsmpp/issues/6#issuecomment-133002430>
> .
>



-- 
Best regards, Mykola Pavluchynskyi

**Comment by bodunov** [comment]
I can easily reproduce the same problem. I am using Camel-Smpp. Create an application with Smpp Consumer, deploy it. It will connect to the server and start sending enquire link pdus. Then stop the server. And the AbstractSession will get stuck on the join in close().

**Comment by rlusnikov** [comment]
I had the same problem. Here is dump from Java Wrapper where threads are blocked:
```
INFO   | jvm 1    | 2016/12/07 09:33:05 | "EnquireLinkSender: org.jsmpp.session.SMPPSession@f706a2f" #25 prio=5 os_prio=0 tid=0x00007f23eca0d000 nid=0x65a1 waiting for monitor entry [0x00007f2430a16000]
INFO   | jvm 1    | 2016/12/07 09:33:05 |    java.lang.Thread.State: BLOCKED (on object monitor)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at org.jsmpp.session.SMPPSessionContext.getSessionState(SMPPSessionContext.java:39)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       - waiting to lock <0x00000000e0bae708> (a org.jsmpp.session.SMPPSessionContext)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at org.jsmpp.session.AbstractSession.getSessionState(AbstractSession.java:114)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at org.jsmpp.session.AbstractSession.isReadPdu(AbstractSession.java:118)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       - locked <0x00000000e0ba4f48> (a org.jsmpp.session.SMPPSession)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at org.jsmpp.session.AbstractSession$EnquireLinkSender.run(AbstractSession.java:461)

INFO   | jvm 1    | 2016/12/07 09:33:05 | "PDUReaderWorker: org.jsmpp.session.SMPPSession@f706a2f" #23 prio=5 os_prio=0 tid=0x00007f23eca08800 nid=0x659d in Object.wait() [0x00007f2430c18000]
INFO   | jvm 1    | 2016/12/07 09:33:05 |    java.lang.Thread.State: WAITING (on object monitor)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at java.lang.Object.wait(Native Method)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at java.lang.Thread.join(Thread.java:1249)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       - locked <0x00000000e0ba50a8> (a org.jsmpp.session.AbstractSession$EnquireLinkSender)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at java.lang.Thread.join(Thread.java:1323)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at org.jsmpp.session.AbstractSession.close(AbstractSession.java:220)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at org.jsmpp.session.AbstractSession.unbindAndClose(AbstractSession.java:402)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at org.apache.camel.component.smpp.SmppProducer.closeSession(SmppProducer.java:170)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at org.apache.camel.component.smpp.SmppProducer.access$200(SmppProducer.java:43)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at org.apache.camel.component.smpp.SmppProducer$1.onStateChange(SmppProducer.java:64)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at org.jsmpp.session.AbstractSessionContext.fireStateChanged(AbstractSessionContext.java:79)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at org.jsmpp.session.SMPPSessionContext.changeState(SMPPSessionContext.java:61)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at org.jsmpp.session.AbstractSessionContext.close(AbstractSessionContext.java:62)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       - locked <0x00000000e0bae708> (a org.jsmpp.session.SMPPSessionContext)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at org.jsmpp.session.AbstractSession.close(AbstractSession.java:207)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at org.jsmpp.session.SMPPSession$PDUReaderWorker.readPDU(SMPPSession.java:623)
INFO   | jvm 1    | 2016/12/07 09:33:05 |       at org.jsmpp.session.SMPPSession$PDUReaderWorker.run(SMPPSession.java:582)
INFO   | jvm 1    | 2016/12/07 09:33:05 |
```
I have solved this problem by adding the counter check:

```java
        // Make sure the enquireLinkThread doesn't wait for itself
        if (Thread.currentThread() != enquireLinkSender) {
            logger.info("Closing enquireLinkSender for session {}", enquireLinkSender, sessionId);
            if (enquireLinkSender != null) {
                int counter = 10; // Deadlock ???
                while(enquireLinkSender.isAlive() && counter-- > 0) {
                    try {
                        enquireLinkSender.join(5000);
                        logger.debug("Closing enquireLinkSender counter = {}", counter);
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                        logger.warn("Interrupted while waiting for enquireLinkSender thread to exit");
                    }
                }
                
                if (counter == 0)
                    enquireLinkSender.interrupt();
            }
        }
```

**Comment by pmoerenhout** [comment]
I will investigate the problem what the exact cause is of the thread problem.

**Comment by rlusnikov** [comment]
All the same in 2.3.3-SNAPSHOT:

```
debug:
11:02:53.549 [pool-4-thread-3] DEBUG org.jsmpp.session.PDUProcessTask - Received SMPP message PDUHeader(16, 00000015, 00000000, 5)
11:02:53.549 [pool-4-thread-3] DEBUG org.jsmpp.session.SMPPSession - Sending enquire_link_resp
11:02:53.549 [pool-4-thread-3] DEBUG org.jsmpp.DefaultPDUSender - Sending SMPP message 00 00 00 10 80 00 00 15 00 00 00 00 00 00 00 05
11:02:58.367 [pool-3-thread-1] DEBUG org.jsmpp.session.PDUProcessTask - Received SMPP message PDUHeader(16, 00000015, 00000000, 6)
11:02:58.367 [pool-3-thread-1] DEBUG org.jsmpp.session.SMPPSession - Sending enquire_link_resp
11:02:58.367 [pool-3-thread-1] DEBUG org.jsmpp.DefaultPDUSender - Sending SMPP message 00 00 00 10 80 00 00 15 00 00 00 00 00 00 00 06
11:02:58.551 [pool-4-thread-1] DEBUG org.jsmpp.session.PDUProcessTask - Received SMPP message PDUHeader(16, 00000015, 00000000, 6)
11:02:58.551 [pool-4-thread-1] DEBUG org.jsmpp.session.SMPPSession - Sending enquire_link_resp
11:02:58.551 [pool-4-thread-1] DEBUG org.jsmpp.DefaultPDUSender - Sending SMPP message 00 00 00 10 80 00 00 15 00 00 00 00 00 00 00 06
11:03:03.571 [PDUReaderWorker: org.jsmpp.session.SMPPSession@137a0b5c] WARN  org.jsmpp.session.SMPPSession - IOException while reading: null
11:03:03.571 [PDUReaderWorker: org.jsmpp.session.SMPPSession@4e793a3] WARN  org.jsmpp.session.SMPPSession - IOException while reading: null
11:03:03.571 [PDUReaderWorker: org.jsmpp.session.SMPPSession@137a0b5c] DEBUG org.jsmpp.session.AbstractSession - Close session 9391d31f
11:03:03.571 [PDUReaderWorker: org.jsmpp.session.SMPPSession@4e793a3] DEBUG org.jsmpp.session.AbstractSession - Close session f4dbe4c4
11:03:03.571 [PDUReaderWorker: org.jsmpp.session.SMPPSession@137a0b5c] WARN  org.apache.camel.component.smpp.SmppProducer - Lost connection to: smpp://beeline@localhost:8057 - trying to reconnect...
11:03:03.571 [PDUReaderWorker: org.jsmpp.session.SMPPSession@4e793a3] WARN  org.apache.camel.component.smpp.SmppConsumer - Lost connection to: smpp://beeline@localhost:8057 - trying to reconnect...
11:03:04.571 [PDUReaderWorker: org.jsmpp.session.SMPPSession@4e793a3] DEBUG org.jsmpp.session.AbstractSession - Unbind and close sesssion f4dbe4c4
11:03:04.571 [PDUReaderWorker: org.jsmpp.session.SMPPSession@137a0b5c] DEBUG org.jsmpp.session.AbstractSession - Unbind and close sesssion 9391d31f
11:03:04.572 [PDUReaderWorker: org.jsmpp.session.SMPPSession@4e793a3] DEBUG org.jsmpp.session.AbstractSession - Close session f4dbe4c4
11:03:04.572 [PDUReaderWorker: org.jsmpp.session.SMPPSession@137a0b5c] DEBUG org.jsmpp.session.AbstractSession - Close session 9391d31f
11:03:04.572 [PDUReaderWorker: org.jsmpp.session.SMPPSession@4e793a3] DEBUG org.jsmpp.session.AbstractSession - Closing enquireLinkSender for session f4dbe4c4
11:03:04.572 [PDUReaderWorker: org.jsmpp.session.SMPPSession@137a0b5c] DEBUG org.jsmpp.session.AbstractSession - Closing enquireLinkSender for session 9391d31f

dump:
INFO   | jvm 1    | 2016/12/12 11:04:05 | "EnquireLinkSender: org.jsmpp.session.SMPPSession@4e793a3" #32 prio=5 os_prio=0 tid=0x00007f0aa4b40800 nid=0x2245 waiting for monitor entry [0x00007f0af1e83000]
INFO   | jvm 1    | 2016/12/12 11:04:05 |    java.lang.Thread.State: BLOCKED (on object monitor)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at org.jsmpp.session.SMPPSessionContext.getSessionState(SMPPSessionContext.java:39)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       - waiting to lock <0x00000000f7e96788> (a org.jsmpp.session.SMPPSessionContext)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at org.jsmpp.session.AbstractSession.getSessionState(AbstractSession.java:114)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at org.jsmpp.session.AbstractSession.isReadPdu(AbstractSession.java:118)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       - locked <0x00000000f7e965c0> (a org.jsmpp.session.SMPPSession)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at org.jsmpp.session.AbstractSession$EnquireLinkSender.run(AbstractSession.java:460)

INFO   | jvm 1    | 2016/12/12 11:04:05 | "PDUReaderWorker: org.jsmpp.session.SMPPSession@4e793a3" #30 prio=5 os_prio=0 tid=0x00007f0aa4b3f800 nid=0x2240 in Object.wait() [0x00007f0af2085000]
INFO   | jvm 1    | 2016/12/12 11:04:05 |    java.lang.Thread.State: WAITING (on object monitor)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at java.lang.Object.wait(Native Method)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       - waiting on <0x00000000f7ea8380> (a org.jsmpp.session.AbstractSession$EnquireLinkSender)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at java.lang.Thread.join(Thread.java:1249)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       - locked <0x00000000f7ea8380> (a org.jsmpp.session.AbstractSession$EnquireLinkSender)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at java.lang.Thread.join(Thread.java:1323)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at org.jsmpp.session.AbstractSession.close(AbstractSession.java:220)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at org.jsmpp.session.AbstractSession.unbindAndClose(AbstractSession.java:401)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at org.apache.camel.component.smpp.SmppConsumer.closeSession(SmppConsumer.java:133)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at org.apache.camel.component.smpp.SmppConsumer.access$200(SmppConsumer.java:45)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at org.apache.camel.component.smpp.SmppConsumer$1.onStateChange(SmppConsumer.java:73)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at org.jsmpp.session.AbstractSessionContext.fireStateChanged(AbstractSessionContext.java:79)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at org.jsmpp.session.SMPPSessionContext.changeState(SMPPSessionContext.java:61)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at org.jsmpp.session.AbstractSessionContext.close(AbstractSessionContext.java:62)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       - locked <0x00000000f7e96788> (a org.jsmpp.session.SMPPSessionContext)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at org.jsmpp.session.AbstractSession.close(AbstractSession.java:207)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at org.jsmpp.session.SMPPSession$PDUReaderWorker.readPDU(SMPPSession.java:623)
INFO   | jvm 1    | 2016/12/12 11:04:05 |       at org.jsmpp.session.SMPPSession$PDUReaderWorker.run(SMPPSession.java:582)
```

**Comment by pmoerenhout** [comment]
I also made some changes to the camel-smpp component. I get the reconnect working, but closing the application still deadlocks. I have to figure out why in Camel the synchronized for the Session objects blocks, and in a plain Java not.

**Comment by pmoerenhout** [comment]
I modified AbstractSession. The EnquireLinkSender thread is now always terminated in Apache Camel. rlusnikov, could you retry the test?

**Comment by pmoerenhout** [comment]
I released jSMPP 2.3.3 which should fix it. I also made some code changes in camel-smpp 2.19.0, to remove a 1 second sleep, when closing the session.
